### PR TITLE
Complete all basic features.

### PR DESCRIPTION
**Pull request by Qubrick**
You should review code now.

### CODE REVIEW

**Comment by Qubrick** [comment]
I use notepad++ as editor. You're right about jquery. Maybe I might change it some way.

File: webAO/client.js [hunk 1]
```diff
@@ -146,6 +247,107 @@ class Client {
...
+		background_select.add(new Option("Custom", 0));
+		for(let i = 1; i <= background_arr.length; i++) {
+		  background_select.add(new Option(background_arr[i - 1]));
+		}
+		// Calculate gif duration of shouts
+		let shouts = ["holdit", "objection", "takethat"];
+		for (let i = 0; i < shouts.length; i++) {
+			let shout_src = AO_HOST + this.resources[shouts[i]]["src"];
+			FileExist(shout_src, this.callbackLoadImageResources, shouts[i]);		
+		}
```
[the remark this answers is not in the document]
Comment (gameboyprinter) [reply webAO/client.js]: the code is easier to read in the regular format imo

File: webAO/client.js [hunk 2]
```diff
@@ -146,6 +247,107 @@ class Client {
...
+		
+		// Calculate gif duration of testimony
+		let testimony = ["witnesstestimony", "crossexamination"];
+		for (let i = 0; i < testimony.length; i++) {
```
[the remark this answers is not in the document]
Comment (gameboyprinter) [reply webAO/client.js]: the code is easier to read in the regular format imo

File: webAO/client.js [hunk 3]
```diff
@@ -378,6 +580,34 @@ class Client {
...
+			}
+		}
+		
+		var evidence_box = document.getElementById("evidences");
+		evidence_box.innerHTML = "";
+		for(let i = 1; i <= this.evidences.length; i++){
+			evidence_box.innerHTML += '<img src="' + this.evidences[i - 1].icon + 
+				'" id="evi_' + i +'" alt="' + this.evidences[i - 1].name +
+				'" class="client_button" ' +
+				'onclick="pickevidence('+ i +')">';
```
[the remark this answers is not in the document]
Comment (gameboyprinter) [reply webAO/client.js]: can't use forEach, as this requires i to be accessible

[the remark this answers is not in the document]
Comment (oldmud0) [reply webAO/client.js]: The index is passed in as the second argument in the forEach lambda.

File: webAO/client.js [hunk 4]
```diff
@@ -641,50 +947,201 @@ class Viewport {
...
+	 * @param {object} param 
+	 */
+	getAnimLength(filename, callback, param) {
+		var request = new XMLHttpRequest();
+		request.open('GET', filename, true);
+		request.responseType = 'arraybuffer';
+		request.addEventListener('load', function () {
+			// Use gify API
+			// https://github.com/rfrench/gify
+			var gifInfo = gify.getInfo(request.response);
```
[the remark this answers is not in the document]
Comment (gameboyprinter) [reply webAO/client.js]: i see nothing wrong with it imo

File: webAO/client.js [hunk 5]
```diff
@@ -641,50 +947,201 @@ class Viewport {
...
 			this.chatmsg.isnew = false;
-			this.chatmsg.startspeaking = true;
+			this.chatmsg.startpreanim = true;
 		}
 
-		if (this.textTimer >= this.shoutTimer) {
-			if (this.chatmsg.startspeaking) {
+		if(this.textTimer >= this.shoutTimer && this.chatmsg.startpreanim) {
+			// Effect stuff
+			if (this.chatmsg.flash == 2){
```
[the remark this answers is not in the document]
Comment (gameboyprinter) [reply webAO/client.js]: is it guaranteed to deserialize as a number? i see nothing wrong with the greedy comparison

File: webAO/client.js [hunk 6]
```diff
@@ -641,50 +947,201 @@ class Viewport {
...
+				//Shake screen
+				this.sfxaudio.pause();
+				this.sfxplayed = 1;
+				this.sfxaudio.src = AO_HOST + "sounds/general/sfx-stab.wav";
+				this.sfxaudio.play();
+				$('#client_gamewindow').effect( "shake",{"direction":"up"});
```
[the remark this answers is not in the document]
Comment (gameboyprinter) [reply webAO/client.js]: this is css, just applied via jquery.

**Pull request merged**